### PR TITLE
/blockchallenges: Support allowing challenges from certain groups

### PR DESCRIPTION
**Pull request by Distrib-ps**
Pr for https://www.smogon.com/forums/threads/allow-blockchallenges-to-specify-which-groups-to-block-them-from.3683632/

### CODE REVIEW

File: server/chat-commands/core.ts [hunk 1]
```diff
@@ -1589,10 +1589,23 @@ export const commands: Chat.ChatCommands = {
...
+			user.settings.blockChallenges = target;
+			target = this.tr(target);
+			this.sendReply(this.tr`You are now blocking challenges, except from staff and ${target} users.`);
+		} else {
+			user.settings.blockChallenges = true;
+			this.sendReply(this.tr`You are now blocking all incoming challenge requests.`);
+		}
 		user.update();
-		this.sendReply(this.tr`You are now blocking all incoming challenge requests.`);
+		return true;
```
Comment (AnnikaCodes) [inline server/chat-commands/core.ts:1608]:
```suggestion
```

File: server/chat.ts [hunk 2]
```diff
@@ -1115,6 +1115,11 @@ export class CommandContext extends MessageContext {
 					targetUser.id !== user.id) {
 					throw new Chat.ErrorMessage(this.tr`You are blocking private messages right now.`);
 				}
+				if (user.settings.blockChallenges && (user.settings.blockChallenges === true ||
+					!Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) && !targetUser.can('bypassblocks') &&
+					targetUser.id !== user.id) {
+					throw new Chat.ErrorMessage(this.tr`You are blocking challenge right now.`);
+				}
```
Comment (AnnikaCodes) [inline server/chat.ts:1122]:
```suggestion
				if (
					user.settings.blockChallenges && 
					(user.settings.blockChallenges === true || !Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) &&
					!targetUser.can('bypassblocks') &&
					targetUser.id !== user.id
				) {
					throw new Chat.ErrorMessage(this.tr`You are blocking challenges right now.`);
				}
```

Comment (AnnikaCodes) [reply server/chat.ts:1122]: Note that "You are blocking challenges right now." will need to be updated in all the translations files.

File: server/ladders.ts [hunk 3]
```diff
@@ -163,7 +163,9 @@ class Ladder extends LadderStore {
 			connection.popup(`You are already challenging someone. Cancel that challenge before challenging someone else.`);
 			return false;
 		}
-		if (targetUser.settings.blockChallenges && !user.can('bypassblocks', targetUser)) {
+		if (targetUser.settings.blockChallenges &&
```
Comment (AnnikaCodes) [inline server/ladders.ts:166]: See my other comment about `if` statement formatting.

Comment (Distrib-ps) [reply server/ladders.ts]: Done.

File: server/ladders.ts [hunk 4]
```diff
@@ -163,7 +163,11 @@ class Ladder extends LadderStore {
 			connection.popup(`You are already challenging someone. Cancel that challenge before challenging someone else.`);
 			return false;
 		}
-		if (targetUser.settings.blockChallenges && !user.can('bypassblocks', targetUser)) {
+		if (
+			targetUser.settings.blockChallenges &&
+			(targetUser.settings.blockChallenges === true ||
+			!Users.globalAuth.atLeast(user, targetUser.settings.blockChallenges)) &&
+			!user.can('bypassblocks') && targetUser.id !== user.id) {
```
Comment (AnnikaCodes) [inline server/ladders.ts:170]:
```suggestion
		if (
			user.settings.blockChallenges &&
			(user.settings.blockChallenges === true || !Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) &&
			!targetUser.can('bypassblocks') &&
			targetUser.id !== user.id
		) {
```

File: server/ladders.ts [hunk 5]
```diff
@@ -163,7 +163,12 @@ class Ladder extends LadderStore {
 			connection.popup(`You are already challenging someone. Cancel that challenge before challenging someone else.`);
 			return false;
 		}
-		if (targetUser.settings.blockChallenges && !user.can('bypassblocks', targetUser)) {
+		if (
+			user.settings.blockChallenges &&
+			(user.settings.blockChallenges === true || !Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) &&
+			!targetUser.can('bypassblocks') &&
+			targetUser.id !== user.id
```
Comment (AnnikaCodes) [inline server/ladders.ts:170]:
```suggestion
			!targetUser.can('bypassblocks')
```
Users can't challenge themselves, anyway.

File: server/chat.ts [hunk 6]
```diff
@@ -1115,6 +1115,14 @@ export class CommandContext extends MessageContext {
 					targetUser.id !== user.id) {
 					throw new Chat.ErrorMessage(this.tr`You are blocking private messages right now.`);
 				}
+				if (
+					user.settings.blockChallenges &&
+					(user.settings.blockChallenges === true || !Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) &&
+					!targetUser.can('bypassblocks') &&
+					targetUser.id !== user.id
+				) {
```
Comment (AnnikaCodes) [inline server/chat.ts:1123]:
```suggestion
					!targetUser.can('bypassblocks')
				) {
```

File: server/ladders.ts [hunk 7]
```diff
@@ -163,7 +163,11 @@ class Ladder extends LadderStore {
 			connection.popup(`You are already challenging someone. Cancel that challenge before challenging someone else.`);
 			return false;
 		}
-		if (targetUser.settings.blockChallenges && !user.can('bypassblocks', targetUser)) {
+		if (
+			user.settings.blockChallenges &&
+			(user.settings.blockChallenges === true || !Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) &&
+			!targetUser.can('bypassblocks')
+		) {
```
Comment (AnnikaCodes) [inline server/ladders.ts:170]:
```suggestion
		if (
			targetUser.settings.blockChallenges &&
			(targetUser.settings.blockChallenges === true || !Users.globalAuth.atLeast(user, targetUser.settings.blockChallenges)) &&
			!user.can('bypassblocks', targetUser)
		) {
```
This is backwards; we want to check if the target user is blocking challenges.

File: server/chat.ts [hunk 8]
```diff
@@ -1115,6 +1115,13 @@ export class CommandContext extends MessageContext {
 					targetUser.id !== user.id) {
 					throw new Chat.ErrorMessage(this.tr`You are blocking private messages right now.`);
 				}
+				if (
+					user.settings.blockChallenges &&
+					(user.settings.blockChallenges === true || !Users.globalAuth.atLeast(targetUser, user.settings.blockChallenges)) &&
+					!targetUser.can('bypassblocks')
+				) {
+					throw new Chat.ErrorMessage(this.tr`You are blocking challenges right now.`);
+				}
```
Comment (AnnikaCodes) [inline server/chat.ts:1124]: What is this code for? It seems like it prevents you from PMing users who can't challenge you?

Comment (Distrib-ps) [reply server/chat.ts:1124]: I don't know why anymore. But I think this code is useless.

Comment (AnnikaCodes) [reply server/chat.ts:1124]: Remove it, then.

File: server/ladders.ts [hunk 9]
```diff
@@ -163,7 +163,11 @@ class Ladder extends LadderStore {
 			connection.popup(`You are already challenging someone. Cancel that challenge before challenging someone else.`);
 			return false;
 		}
-		if (targetUser.settings.blockChallenges && !user.can('bypassblocks', targetUser)) {
+		if (
+			targetUser.settings.blockChallenges &&
+			(targetUser.settings.blockChallenges === true || !Users.globalAuth.atLeast(user, targetUser.settings.blockChallenges)) &&
+			!user.can('bypassblocks', targetUser)
+		) {
```
Comment (AnnikaCodes) [inline server/ladders.ts:170]:
```suggestion
		if (targetUser.settings.blockChallenges && !user.can('bypassblocks', targetUser) && (
			targetUser.settings.blockChallenges === true ||
			!Users.globalAuth.atLeast(user, targetUser.settings.blockChallenges)
		)) {
```

**Review by AnnikaCodes — APPROVED** [review]
Looks good, thanks!

**Pull request merged**